src/Data/List/Lazy.purs

```diff
@@ -757,6 +759,17 @@ foldrLazy op z = go
       Cons x xs' -> Z.defer \_ -> x `op` go xs'
       Nil -> z
 
+-- | Perform a right fold until the list is consumed or the provided function returns a Left value
+foldrWhile :: forall a b. (a -> b -> Either b b) -> b -> List a -> b
+foldrWhile f acc xs = go $ step xs
+  where
+    go :: Step a -> b
+    go Nil = acc
+    go (Cons x xs') =
+      case f x acc of
+        Left acc' -> acc'
+        Right acc' -> foldrWhile f acc' xs'
```

Comment:
Did you intend to use `go` here? Otherwise there's no reason for the binding, you can just case on step.

Reply:
Come to think of it, I think using `go` here is probably sensible as we can get TCO without having to thread `f` through for each iteration.

Reply:
I didn't intend to use `go`. The recursive call needs to call `step` again, so it needs to go back in through the top. I'm happy to remove go completely, though the extraction makes it a bit easier for me to read the implementation.

Reply:
This needs to be set up in such a way that TCO fires or it will overflow for long lists in the case where the function keeps returning Right for long enough. I’d suggest:

```purescript
foldlWhile f = go
  where
  go acc xs =
    case step xs of
      Nil ->
        acc
      Cons x ys ->
        case f acc y of
          Left acc’ -> acc’
          Right acc’ -> go acc’ ys
```

Using a binding for `go` like this means that we don’t need to thread `f` back through each time.

Reply:
@hdgarrood for my edification, why wasn't the initial implementation tail recursive? Is it the mutually recursive call from `go` back to `foldlWhile`?

Reply:
I think your implementation could also have been described as tail-recursive, although I was concerned that introducing the `go` binding and then recursing with a call to the outer binding would mean it was tail-recursive in a way that the compiler might not be able to recognise and optimise, if that makes sense?